src/Lib.P2P/Routing/DistributedQuery.cs

```diff
@@ -0,0 +1,327 @@
+            var waits = 20;
+            while (!_runningQuery.IsCancellationRequested && waits > 0)
+            {
+                // Get the nearest peer that has not been visited.
+                var peer = Dht.RoutingTable
+                   .NearestPeers(QueryKey)
+                   .FirstOrDefault(p => !_visited.ContainsKey(p));
+                if (peer == null)
+                {
+                    --waits;
+                    await Task.Delay(100);
+                    continue;
+                }
+
+                if (!_visited.TryAdd(peer, peer))
+                {
+                    continue;
+                }
+                
+                ++pass;
+
+                // Ask the nearest peer.
+                await AskCount.WaitAsync(_runningQuery.Token).ConfigureAwait(false);
+                var start = DateTime.Now;
+                Log.Debug($"Q{Id}.{taskId}.{pass} ask {peer}");
+                try
+                {
+                    using (var timeout = new CancellationTokenSource(AskTime))
+                    {
+                        using var cts = CancellationTokenSource.CreateLinkedTokenSource(timeout.Token, _runningQuery.Token);
+                        {
+                            await using (var stream =
+                                await Dht.SwarmService.DialAsync(peer, Dht.ToString(), cts.Token).ConfigureAwait(false))
+                            {
+                                // Send the KAD query and get a response.
+                                Serializer.SerializeWithLengthPrefix(stream, _queryMessage, PrefixStyle.Base128);
+                                await stream.FlushAsync(cts.Token).ConfigureAwait(false);
+                                var response = await ProtoBufHelper.ReadMessageAsync<DhtMessage>(stream, cts.Token)
+                                   .ConfigureAwait(false);
+
+                                // Process answer
+                                ProcessProviders(response.ProviderPeers);
+                                ProcessCloserPeers(response.CloserPeers);
+                            }
+                        }
+                    }
+
+                    var time = DateTime.Now - start;
+                    Log.Debug($"Q{Id}.{taskId}.{pass} ok {peer} ({time.TotalMilliseconds} ms)");
+                }
+                catch (Exception e)
+                {
+                    Interlocked.Increment(ref _failedConnects);
+                    var time = DateTime.Now - start;
+                    Log.Warn($"Q{Id}.{taskId}.{pass} failed ({time.TotalMilliseconds} ms) - {e.Message}");
+                }
+                finally
+                {
+                    AskCount.Release();
+                }
+            }
+        }
+
+        private void ProcessProviders(DhtPeerMessage[] providers)
+        {
+            if (providers == null)
+            {
+                return;
+            }
+
+            foreach (var provider in providers)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'foreach' block.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)